src/Stack/Solver.hs

```diff
@@ -468,12 +469,20 @@ checkResolverSpec gpds flags resolver = do
       -- TODO support custom resolver for stack init
       ResolverCustom {} -> return $ BuildPlanCheckPartial Map.empty Map.empty
 
--- | Finds all files with a .cabal extension under a given directory.
+-- | Finds all files with a .cabal extension under a given directory. If
+-- a `hpack` `package.yaml` file exists, this will be used to generate a cabal
+-- file.
 -- Subdirectories can be included depending on the @recurse@ parameter.
 findCabalFiles :: MonadIO m => Bool -> Path Abs Dir -> m [Path Abs File]
-findCabalFiles recurse dir =
-    liftIO $ findFiles dir isCabal (\subdir -> recurse && not (isIgnored subdir))
+findCabalFiles recurse dir = do
```

Comment:
We can rename findCabalFiles  to something like findProjectFiles and let it take a predicate like (== "package.yaml") or (".cabal" `isSuffixOf`). Then we can call it once for finding package.yaml in the first step (and then run hpack on the result) and then again for finding .cabal files in the next step. That will cleanly separate these two steps, making the code a tiny bit easier to read.